New Shop data storage system & bug fixes

KillerOfPie opened the pull request:
Xanth3r is going to test some more for me but im pretty sure this is well flushed out now, give it a look and let me know if you have questions.

[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Template

### Pull Request

Replace this sentence with general description of what your Pull Request does.

KillerOfPie committed:
Moved from Storage naming to using Persistent Data Types, minor Reformats and fixes

KillerOfPie committed:
Better Double chest handling

KillerOfPie committed:
Fix for #73

KillerOfPie commented:
this should fix #70

KillerOfPie committed:
Fix for hoppers after removing chest names

KillerOfPie commented:
Everything should be tested and good now

KillerOfPie committed:
Fix for GD claims and NPE on shop signs without shops

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -236,69 +208,52 @@ public void onChestOpen(PlayerInteractEvent e) {
-                return;
-            }
-
-            new ShopChest(block, shopOtherHalf.getOwner(), shopOtherHalf.getShopSign().getLocation()).setEventName(event);
```
why exactly do we not need this anymore?

KillerOfPie replied:
I dont know why this was removed, will add back in when fixing the protection bugs

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:
```diff
@@ -69,8 +68,13 @@ public void onBlockInteract(PlayerInteractEvent e) {
+            s.setLine(0, "");
+            s.setLine(1, "");
+            s.setLine(2, "");
+            s.setLine(3, "");
```
why is the sign being modified here? Also, why is `Sign#update()` not being called after sign modification?

KillerOfPie replied:
If a shop doesnt exist I am clearing the sign since it isnt linked to a shop, I did forget to update the sign tho, will add that in

SparklingComet replied:
mhm.. and under which circumstances would this

```
if (ShopType.isShop(e.getClickedBlock())) {
            s = (Sign) e.getClickedBlock().getState();
        }
```

run, and yet the sign not be connected to a valid shop?

KillerOfPie replied:
XANT3R had an error come up where the shop sign existed but there was no shop in the config, which this prevents, should also clear the sign to prevent it from continuing to happen

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -68,36 +68,82 @@ public ShopChest(Block chest, UUID owner, Location sign) {
+            if (isDoubleChest(checking)) {
+                DoubleChest dbl = getDoubleChest(checking);
+                return ((Container) dbl.getLeftSide().getInventory().getLocation().getBlock()).getPersistentDataContainer().has(plugin.getStorageKey(), PersistentDataType.STRING) ||
+                        ((Container) dbl.getRightSide().getInventory().getLocation().getBlock()).getPersistentDataContainer().has(plugin.getStorageKey(), PersistentDataType.STRING);
```
ingenious use of PDC, I guess

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -68,36 +68,82 @@ public ShopChest(Block chest, UUID owner, Location sign) {
+                    bs.update();
+                }
+            }
+        }
```
what exactly is happening here?

KillerOfPie replied:
clearing out our old name from any block that is passed to this, if it's a double chest it clears both sides.

KillerOfPie committed:
Fix for NPE when iTrade does not have a shop

SparklingComet commented on src/main/java/org/shanerx/tradeshop/TradeShop.java:
```diff
@@ -42,6 +43,18 @@
 
+
+	private NamespacedKey storageKey = new NamespacedKey(this, "tradeshop-storage-data");
+	private NamespacedKey signKey = new NamespacedKey(this, "tradeshop-sign-data");
```
just technicality, make `final` maybe?

KillerOfPie replied:
yes, will do

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/CustomInventoryListener.java:
```diff
@@ -27,14 +27,15 @@
 public class CustomInventoryListener extends Utils implements Listener {
 
-	@EventHandler
+    @EventHandler(priority = EventPriority.HIGHEST)
```
I know the `ignoreCancelled` flag does not seem to be working, but did you not put it here on purpose?

KillerOfPie replied:
yes because i do want to check it no matter what since its checking for our own inventory

SparklingComet replied:
I guess, but what if some other plugin has anything to say about it?

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -144,79 +142,53 @@ public void onEntityExplodeItem(EntityExplodeEvent event) {
-                new ShopChest(block.getLocation()).resetName();
+        } else if (ShopChest.isShopChest(block)) {
+            shop = new ShopChest(block.getLocation()).getShop();
+            if (shop == null)
```
sleek

KillerOfPie replied:
this was added a bit ago, just didn't get it everywhere

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -144,79 +142,53 @@ public void onEntityExplodeItem(EntityExplodeEvent event) {
+                }
 
-            Shop shop = Shop.loadShop(s);
+                shop.getChestAsSC().resetName();
```
sorry, but what exactly does this do? Remove the old chest name `$ ^Sign:...`, or something else?

KillerOfPie replied:
yes, removes the old chestnames

SparklingComet replied:
are names of other (shop-)chests reset as well or only the ones that are broken?

KillerOfPie replied:
This is in a couple places including the interact listener, so any shop that is interacted with should be reset if it still has one of our names

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -144,79 +142,53 @@ public void onEntityExplodeItem(EntityExplodeEvent event) {
-
-                        bs.update();
-                    }
+                    shop.updateSign();
```
what exactly is this for?

KillerOfPie replied:
updates the text on the sign when the chest is broken, including status and everything

SparklingComet replied:
ty for clarifying

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -157,48 +203,42 @@ public String getName() {
+			container2.getPersistentDataContainer().set(plugin.getStorageKey(), PersistentDataType.STRING, getName());
+			container2.update();
 		}
+*/
```
why is the dchest part gone?

KillerOfPie replied:
Wasn't sure if i needed to add the item to both sides of a double chest

SparklingComet reviewed (commented):
The changes so far are very good. Once the few bugs are fixed, this will make for a very good update.

KillerOfPie reviewed (commented):
@SparklingComet Resolve any items that I gave a sufficient answer on.

KillerOfPie committed:
Bug Fixes and Review Changes
- CommandRunner
  * Updated open command to use the additional shop statuses, as well as give more accurate messages
  * Fixed NPEs coming from #findShop
- ShopTradeListener
  * Added sign update after clear
- Shop
  * Updated #setOpen to return the new status
  * #isMissingItems now only checks cost if the shop is an iTrade
  * Fixed #updateStatus so that iTrade shops will be set to OPEN and never OUT_OF_STOCK
- TradeShop
  * Made NamespacedKey objects final

KillerOfPie committed:
Minor Changes to Build and API Ver
- plugin.yml
  * Reverted API-Version to 1.14
  * changed version, description, and website to maven project variables, only need to change things in the pom now
- pom.xml
  * moved some things around
  * changed url variable
  * added description variable and packaging
  * added source directory and moved resources to top

KillerOfPie committed:
Bug Fixes and Review Changes
- CommandRunner
  * Fixed Remove and List commands to start from 1
- Shop
  * Fixed Cost being required for iTrade when it should be Product
- Utils
  * Removed unnecessary set on init in #canExchangeAll
  * Fix for error when using iTrade with only product

KillerOfPie committed:
Minor Changes to Build and API Ver
- PlayerShopDestroyEvent
  * Added destroyBlock boolean and getter/setter
- JoinEventListener
  * Removed unneeded ignoreCancelled
- ShopCreateListener, ShopProtectionListener
  * Added returns if event#isCancelled
- ShopRestockListener
  * Removed unneeded ignoreCancelled
- ShopTradeListener
  * Added return if event#isCancelled or UseInteractedBlock is Denied
- Updater
  * Added Test toBuildtype enums

KillerOfPie committed:
Potential Fix for comparing Shulker inventories during trade
- Utils
  * Added #compareShulkers
  * implemented #compareShulkers in #getItems

KillerOfPie committed:
Bug Fixes
- ShopProtectionListener
  * Fix for #80 and #82
- Utils
  * Fix for #81

SparklingComet committed:
Fix NPE when console issues player-only commands

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -910,8 +910,12 @@ public void multi() {
+			sendMessage(Message.PLAYER_ONLY_COMMAND.getPrefixed());
+			return null;
+		}
 
```
This fixes #84 and similar NPEs for other player-only subcommands

KillerOfPie committed:
Changed POM Dependency to MC 1.16.2

KillerOfPie committed:
Changed Version to 2.2.2-DEV

KillerOfPie committed:
Minor Bugs fixes

KillerOfPie committed:
Minor Bugs fixes

KillerOfPie committed:
Minor Bugs fixes

KillerOfPie committed:
Cleanup Message Enums

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopCreateListener.java:
```diff
@@ -124,17 +123,11 @@ public void onSignChange(SignChangeEvent event) {
 				cost = lineCheck(event.getLine(2));
 
-		if (product != null)
+        if (product != null && shop.getProduct().size() == 0)
```
Why `==0`?

KillerOfPie replied:
Because its checking if the array is  empty. I think #isEmpty() may be an alternative same with <1 but they all kinda do the same thing...

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopCreateListener.java:
```diff
@@ -124,17 +123,11 @@ public void onSignChange(SignChangeEvent event) {
 			shop.setProduct(product);
 
-		if (cost != null)
+        if (cost != null && shop.getCost().size() == 0)
```
Same Q here

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -55,8 +55,7 @@
 
-
-		getBlock();
+        getBlock();
```
Indents are all wonky

KillerOfPie committed:
Fix for none trading bug

KillerOfPie committed:
More Fixes

KillerOfPie committed:
Config Comment Support and Min/Max settings for Multi
- Added support for commented settings config, adds comments from code on save of the file
- Added config fix sections for any items that changed paths
- Config Ver incremented to 1.1(these are only to dictate changes and quickly detect what updates need to be made, so all increments will be +0.1)
- Added Min/Max settings for Multi command
- Added Feature Disabled message for multi when the command is run
- Added resource YML t view layout before moving to code

KillerOfPie committed:
Messages Config Comment Support
- Copied new code from settings
- Only comment for now is the Header, can add better sorting and comments later

KillerOfPie committed:
Minor Fix/Changes and Message/Setting Comments
- CommandRunner
  * Changed MULTI_TRADE_DEFAULT check/set back to 2 as the default is not the minimum
- Utils
  * Changed Failed Sign 1st line from ChatColor to Setting value
- Setting
  * Fully commented and sorted
- Message
  * Found and Marked unused values, will not save to file
  * Commented values that had previous comments on GitHub wiki configuration help
  * Needs to be full commented and sectioned for readability

KillerOfPie committed:
Added support for User Purchasing

KillerOfPie committed:
Changing Item Comparison + Other minor tweaks
- Changed some hard coded text to configurable
- Added ShopItemStack to hold Cost/Product Items, handle adjustable configuration, and convert to B64 while being serializable
- Updates bstats to 1.7, now using bstats Plugin ID

KillerOfPie committed:
Forgot to add ShopItemStack

KillerOfPie committed:
Merge remote-tracking branch 'origin/hidingchestname' into hidingchestname

KillerOfPie committed:
Cleanup and added Serialize/Deserialize for ShopItemStack

KillerOfPie committed:
Added Dependecy for better gui menus and fix for unbreakable iTrade
- https://github.com/Phoenix616/InventoryGui/tree/master <- dependency for GUI menus

KillerOfPie committed:
Small fixes and changes, Server is hanging when signed books are used

KillerOfPie committed:
Changes and fixes to get the new comparison system working.

KillerOfPie committed:
Fix for NPE when fixing old shop data

KillerOfPie committed:
Fix for NPE when fixing old shop data

KillerOfPie committed:
Fix for bug where shop doesnt see restocked inventory

KillerOfPie committed:
2.2.2-STABLE Release
